File: ai_companions_flutter/lib/core/services/chat_service.dart
// Chat Service with SSE Streaming Support

import 'dart:async';
import 'dart:convert';
import 'package:dio/dio.dart';
import '../../core/network/dio_client.dart';
import '../../core/config/app_config.dart';
import '../../shared/models/index.dart';

class ChatService {
  ChatService._();

  static final instance = ChatService._();

  Dio get _dio => DioClient.instance.dio;

  Future<List<Conversation>> listConversations() async {
    try {
      final response = await _dio.get('/chat/conversations');
      final data = response.data['data'] as List;
      return data.map((e) => Conversation.fromJson(e as Map<String, dynamic>)).toList();
    } catch (e) {
      rethrow;
    }
  }

  Future<Conversation> createConversation(String characterId) async {
    try {
      final response = await _dio.post(
        '/chat/conversations',
        data: {'character_id': characterId},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return Conversation.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<ConversationData> getConversation(String id, {int page = 1, int limit = 50}) async {
    try {
      final response = await _dio.get(
        '/chat/conversations/$id',
        queryParameters: {'page': page, 'limit': limit},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return ConversationData.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> deleteConversation(String id) async {
    try {
      await _dio.delete('/chat/conversations/$id');
    } catch (e) {
      rethrow;
    }
  }

  Future<SendMessageResponse> sendMessage(String conversationId, String content) async {
    try {
      final response = await _dio.post(
        '/chat/conversations/$conversationId/messages',
        data: {'content': content},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return SendMessageResponse.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> streamMessage(
    String conversationId,
    String content, {
    required void Function(Message) onUserMessage,
    required void Function(String) onToken,
    required void Function(Message) onDone,
    required void Function(String) onError,
  }) async {
    try {
      final url = '${AppConfig.apiBaseUrl}/chat/conversations/$conversationId/stream';

      final response = await _dio.post<ResponseBody>(
        url,
        data: {'content': content},
        options: Options(
          responseType: ResponseType.stream,
          headers: {'Accept': 'text/event-stream'},
        ),
      );

      final completer = Completer<void>();
      String buffer = '';
      String currentEvent = '';

      response.data!.stream.listen(
        (data) {
          buffer += utf8.decode(data);
          final lines = buffer.split('\n');
          buffer = lines.last;

          for (int i = 0; i < lines.length - 1; i++) {
            final line = lines[i].trim();
            if (line.isEmpty) continue;

            if (line.startsWith('event: ')) {
              currentEvent = line.substring(7).trim();
            } else if (line.startsWith('data: ')) {
              final rawData = line.substring(6).trim();
              try {
                final json = jsonDecode(rawData);
                switch (currentEvent) {
                  case 'user_message':
                    onUserMessage(Message.fromJson(json));
                    break;
                  case 'token':
                    onToken(json['content'] ?? '');
                    break;
                  case 'done':
                    onDone(Message.fromJson(json));
                    break;
                  case 'error':
                    onError(json['error'] ?? 'Unknown error');
                    break;
                }
              } catch (e) {
                // Ignore JSON parse errors
              }
            }
          }
        },
        onDone: () => completer.complete(),
        onError: (e) {
          onError('Stream error');
          completer.completeError(e);
        },
      );

      await completer.future;
    } catch (e) {
      rethrow;
    }
  }

  Future<RegenerateResponse> regenerate(String conversationId) async {
    try {
      final response = await _dio.post('/chat/conversations/$conversationId/regenerate');
      final data = response.data['data'] as Map<String, dynamic>;
      return RegenerateResponse.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<SendGiftResponse> sendGift(String conversationId, String giftId) async {
    try {
      final response = await _dio.post(
        '/chat/conversations/$conversationId/gift',
        data: {'gift_id': giftId},
      );
      final data = response.data['data'] as Map<String, dynamic>;
      return SendGiftResponse.fromJson(data);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> deleteMessage(String conversationId, String messageId) async {
    try {
      await _dio.delete('/chat/conversations/$conversationId/messages/$messageId');
    } catch (e) {
      rethrow;
    }
  }
}

File: app/src/services/chat.service.ts
import { api } from './api';
import { tokenStorage } from './tokenStorage';
import type { Conversation, Message, SendMessageRequest } from '@ai-companions/shared';

const API_BASE = process.env.EXPO_PUBLIC_API_URL || 'http://localhost:3001/api/v1';

export const chatService = {
  async listConversations(): Promise<Conversation[]> {
    const response = await api.get('/chat/conversations');
    return response.data.data;
  },

  async createConversation(characterId: string): Promise<Conversation> {
    const response = await api.post('/chat/conversations', { character_id: characterId });
    return response.data.data;
  },

  async getConversation(id: string, page: number = 1, limit: number = 50) {
    const response = await api.get(`/chat/conversations/${id}`, { params: { page, limit } });
    return response.data.data;
  },

  async deleteConversation(id: string): Promise<void> {
    await api.delete(`/chat/conversations/${id}`);
  },

  async sendMessage(conversationId: string, content: string): Promise<{ user_message: Message; ai_message: Message }> {
    const response = await api.post(`/chat/conversations/${conversationId}/messages`, { content });
    return response.data.data;
  },

  /**
   * Stream a message via SSE — calls onToken for each chunk, onDone when complete
   */
  async streamMessage(
    conversationId: string,
    content: string,
    callbacks: {
      onUserMessage: (msg: Message) => void;
      onToken: (token: string) => void;
      onDone: (aiMessage: Message) => void;
      onError: (error: string) => void;
    },
  ): Promise<void> {
    const token = await tokenStorage.getItem('access_token');
    const url = `${API_BASE}/chat/conversations/${conversationId}/stream`;

    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.open('POST', url);
      xhr.setRequestHeader('Content-Type', 'application/json');
      xhr.setRequestHeader('Authorization', `Bearer ${token}`);
      
      let lastProcessedIndex = 0;
      let buffer = '';
      let currentEvent = '';

      xhr.onreadystatechange = () => {
        if (xhr.readyState === 3 || xhr.readyState === 4) {
          const newData = xhr.responseText.slice(lastProcessedIndex);
          lastProcessedIndex = xhr.responseText.length;
          buffer += newData;

          const lines = buffer.split('\n');
          buffer = lines.pop() || '';

          for (const line of lines) {
            const trimmed = line.trim();
            if (!trimmed) {
              currentEvent = '';
              continue;
            }

            if (trimmed.startsWith('event: ')) {
              currentEvent = trimmed.slice(7).trim();
            } else if (trimmed.startsWith('data: ')) {
              if (!currentEvent) continue;
              const rawData = trimmed.slice(6).trim();
              try {
                const data = JSON.parse(rawData);
                switch (currentEvent) {
                  case 'user_message': callbacks.onUserMessage(data); break;
                  case 'token': callbacks.onToken(data.content); break;
                  case 'done': callbacks.onDone(data); break;
                  case 'error': callbacks.onError(data.error); break;
                }
              } catch (e) {
                // Ignore chunked JSON fragments
              }
            }
          }
        }

        if (xhr.readyState === 4) {
          if (xhr.status >= 200 && xhr.status < 300) {
            resolve();
          } else {
            const error = xhr.responseText || `HTTP ${xhr.status}`;
            callbacks.onError(error);
            reject(new Error(error));
          }
        }
      };

      xhr.onerror = () => {
        callbacks.onError('Network request failed');
        reject(new Error('Network request failed'));
      };

      xhr.send(JSON.stringify({ content }));
    });
  },

  async regenerate(conversationId: string): Promise<{ ai_message: Message }> {
    const response = await api.post(`/chat/conversations/${conversationId}/regenerate`);
    return response.data.data;
  },

  async sendGift(conversationId: string, giftId: string): Promise<{
    gift_message: Message;
    ai_message: Message;
    gift: any;
    new_balance: number;
  }> {
    const response = await api.post(`/chat/conversations/${conversationId}/gift`, { gift_id: giftId });
    return response.data.data;
  },

  async deleteMessage(conversationId: string, messageId: string): Promise<void> {
    await api.delete(`/chat/conversations/${conversationId}/messages/${messageId}`);
  },
};
